middleware/rate_limiter.go

```diff
@@ -0,0 +1,97 @@
+package middleware
+
+import (
+	"github.com/labstack/echo/v4"
+	"golang.org/x/time/rate"
+	"net/http"
+	"sync"
+)
+
+// TokenStore is the interface to be implemented by custom stores.
+type TokenStore interface {
+	// Stores for the rate limiter have to implement the ShouldAllow method
+	ShouldAllow(identifier string) bool
+}
+
+type (
+	// RateLimiterConfig defines the configuration for the rate limiter
+	RateLimiterConfig struct {
+		Skipper    Skipper
+		BeforeFunc BeforeFunc
+		// SourceFunc uses echo.Context to extract the identifier for a visitor
+		SourceFunc func(context echo.Context) string
+		// Store defines a store for the rate limiter
+		Store TokenStore
+	}
+)
+
+// DefaultRateLimiterConfig defines default values for RateLimiterConfig
+var DefaultRateLimiterConfig = RateLimiterConfig{
+	Skipper: DefaultSkipper,
+}
+
+// RateLimiter returns a rate limiting middleware
+func RateLimiter(source func(context echo.Context) string, store TokenStore) echo.MiddlewareFunc {
+	config := DefaultRateLimiterConfig
+	config.SourceFunc = source
+	config.Store = store
+
+	return RateLimiterWithConfig(config)
+}
+
+// RateLimiterWithConfig returns a rate limiting middleware
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if config.SourceFunc == nil {
+		panic("Source function must be provided")
+	}
+	if config.Store == nil {
+		panic("Store configuration must be provided")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			if config.BeforeFunc != nil {
+				config.BeforeFunc(c)
+			}
+
+			identifier := config.SourceFunc(c)
+
+			allowed := config.Store.ShouldAllow(identifier)
+			if !allowed {
+				return c.JSON(http.StatusTooManyRequests, nil)
+			}
+			return next(c)
+		}
+	}
+}
+
+// InMemoryStore is the built-in store implementation for RateLimiter
+type InMemoryStore struct {
```

Comment:
As the store is also very specific for the rate limiter the naming should be adjusted too probably.
Maybe `RateLimiterMemoryStore`?

Reply:
Will do this too!